[Bug]: Goroutine Leaks in NewDockerCompose

### Testcontainers version

0.26.0

### Using the latest Testcontainers version?

Yes

### Host OS

Linux

### Host arch

x86_64

### Go version

1.21.3

### Docker version

```shell
Client:
 Version:           20.10.5+dfsg1
 API version:       1.41
 Go version:        go1.15.15
 Git commit:        55c4c88
 Built:             Mon May 30 18:34:49 2022
 OS/Arch:           linux/amd64
 Context:           default
 Experimental:      true

Server:
 Engine:
  Version:          20.10.5+dfsg1
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.15.15
  Git commit:       363e9a8
  Built:            Mon May 30 18:34:49 2022
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.4.13~ds1
  GitCommit:        1.4.13~ds1-1~deb11u4
 runc:
  Version:          1.0.0~rc93+ds1
  GitCommit:        1.0.0~rc93+ds1-5+deb11u2
 docker-init:
  Version:          0.19.0
  GitCommit:
```


### Docker info

```shell
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc., v0.11.2)
  compose: Docker Compose (Docker Inc., v2.20.2)

Server:
 Containers: 0
  Running: 0
  Paused: 0
  Stopped: 0
 Images: 6714
 Server Version: 20.10.5+dfsg1
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
 Logging Driver: json-file
 Cgroup Driver: systemd
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 io.containerd.runtime.v1.linux runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 1.4.13~ds1-1~deb11u4
 runc version: 1.0.0~rc93+ds1-5+deb11u2
 init version: 
 Security Options:
  apparmor
  seccomp
   Profile: default
  cgroupns
 Kernel Version: 5.10.0-25-amd64
 Operating System: Debian GNU/Linux 11 (bullseye)
 OSType: linux
 Architecture: x86_64
 CPUs: 8
 Total Memory: 39.14GiB
 Name: vdi-dd1ah2-158
 ID: PPEH:OYUB:PB3G:EOCK:BISM:UIXT:K4Y3:QMVY:PY2U:7SUR:GOFB:SOME
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
 Default Address Pools:
   Base: 192.168.0.0/16, Size: 24
```


### What happened?

This is somewhat of a follow up to #2007 (possibly same issue, but I'll let the team here reorganize as needed). I did some searching in the current issues and could not find anything on go leaks. 

#### Problem at Hand
Running a simple test that takes up a `NewDockerCompose` from `testcontainers-go/modules/compose` leaks 2 goroutines (as caught by `goleak.VerifyNone(t)`. There are 2 leaks with a wait strategy; without a wait strategy, 4 go routines are leaking. The container or docker compose file does not affect anything. 

Interestingly enough, these problems do not happen if you use the deprecated `NewLocalDockerCompose`. 

I'll include reproduction code below and under "Relevant log output", I'll put the two goroutine leaks you get when you have a wait strategy.

#### What I've Observed/Debugging
It seems to be like 2 of the goroutines come from the creation of the Docker compose and another 2 come from some sort of waiting that happens after. From debugging, it seems like some of those leaks could come from calls to `NewClientHTTPS` or `NewDockerClientWithOpts`. 

Let me know if I can share anything else to help.

### Relevant log output

```shell
--- FAIL: TestIntegrationSuite (1.80s)
    /glnxa64/golang/src/internal/registrystore/redis/store_test.go:50: found unexpected goroutines:
        [Goroutine 52 in state IO wait, with internal/poll.runtime_pollWait on top of the stack:
        goroutine 52 [IO wait]:
        internal/poll.runtime_pollWait(0x7f17fe4dede8, 0x72)
            /glnxa64/golang/src/runtime/netpoll.go:343 +0x85
        internal/poll.(*pollDesc).wait(0xc000714b00?, 0xc00023b000?, 0x0)
            /glnxa64/golang/src/internal/poll/fd_poll_runtime.go:84 +0x27
        internal/poll.(*pollDesc).waitRead(...)
            /glnxa64/golang/src/internal/poll/fd_poll_runtime.go:89
        internal/poll.(*FD).Read(0xc000714b00, {0xc00023b000, 0x1000, 0x1000})
            /glnxa64/golang/src/internal/poll/fd_unix.go:164 +0x27a
        net.(*netFD).Read(0xc000714b00, {0xc00023b000?, 0xc000549ec8?, 0xc000518960?})
            /glnxa64/golang/src/net/fd_posix.go:55 +0x25
        net.(*conn).Read(0xc0000906c8, {0xc00023b000?, 0x0?, 0x0?})
            /glnxa64/golang/src/net/net.go:179 +0x45
        net/http.(*persistConn).Read(0xc0006517a0, {0xc00023b000?, 0x451a20?, 0xc000549ec8?})
            /glnxa64/golang/src/net/http/transport.go:1954 +0x4a
        bufio.(*Reader).fill(0xc0000b6e40)
            /glnxa64/golang/src/bufio/bufio.go:113 +0x103
        bufio.(*Reader).Peek(0xc0000b6e40, 0x1)
            /glnxa64/golang/src/bufio/bufio.go:151 +0x53
        net/http.(*persistConn).readLoop(0xc0006517a0)
            /glnxa64/golang/src/net/http/transport.go:2118 +0x1b9
        created by net/http.(*Transport).dialConn in goroutine 51
            /glnxa64/golang/src/net/http/transport.go:1776 +0x169f
        
         Goroutine 53 in state select, with net/http.(*persistConn).writeLoop on top of the stack:
        goroutine 53 [select]:
        net/http.(*persistConn).writeLoop(0xc0006517a0)
            /glnxa64/golang/src/net/http/transport.go:2421 +0xe5
        created by net/http.(*Transport).dialConn in goroutine 51
            /glnxa64/golang/src/net/http/transport.go:1777 +0x16f1
        ]
FAIL
```


### Additional information

Here's some code that causes the 2 goroutine leaks(docker compose file/wait strategy can be anything, I'm using a Redis container):

```
func TestIntegrationSuite(t *testing.T) {
	defer goleak.VerifyNone(t)
	compose, _ := tc.NewDockerCompose(dockerComposeFile)
	compose.
		WaitForService("redis", wait.NewLogStrategy("Ready to accept connections tcp")). 
		Up(context.Background(), tc.Wait(true))

	t.Cleanup(func() {
		compose.Down(context.Background(), tc.RemoveOrphans(true), tc.RemoveVolumes(true))

	})
```
To see the case where there are 4 leaking goroutines, you just need to comment out the "WaitForService" line.

Here's the code that has no goroutine leaks:
```
func TestIntegrationSuite(t *testing.T) {
	defer goleak.VerifyNone(t)
	identifier := "marisa"

	compose := tc.NewLocalDockerCompose([]string{dockerComposeFile}, identifier)
	err := compose.WithCommand([]string{"up", "-d"}).
		WithExposedService("redis", 6379, wait.ForLog("Ready to accept connections tcp")).
		Invoke()

	if err.Error != nil {
		defer compose.Down()
		t.Fatal("Invoking docker-compose failed with error: ", err)
	}

	t.Cleanup(func() {
		defer compose.Down()
	})
}
```

--- a/modules/compose/compose.go
+++ b/modules/compose/compose.go
@@ -182,6 +182,7 @@ func NewDockerComposeWith(opts ...ComposeStackOption) (*DockerCompose, error) {
 		logger:           composeOptions.Logger,
 		projectProfiles:  composeOptions.Profiles,
 		composeService:   composeService,
+		dockerCli:        dockerCli,
 		dockerClient:     dockerClient,
 		waitStrategies:   make(map[string]wait.Strategy),
 		containers:       make(map[string]*testcontainers.DockerContainer),

--- a/modules/compose/compose_api.go
+++ b/modules/compose/compose_api.go
@@ -15,6 +15,7 @@ import (
 
 	"github.com/compose-spec/compose-go/v2/cli"
 	"github.com/compose-spec/compose-go/v2/types"
+	"github.com/docker/cli/cli/command"
 	"github.com/docker/compose/v5/pkg/api"
 	"github.com/moby/moby/client"
 	"golang.org/x/sync/errgroup"
@@ -180,6 +181,10 @@ type DockerCompose struct {
 	// used to synchronize operations
 	lock sync.RWMutex
 
+	// dockerCli is the Docker CLI instance used internally by the compose service.
+	// It is stored here so its HTTP transport connections can be closed after Down().
+	dockerCli *command.DockerCli
+
 	// name/identifier of the stack that will be started
 	// by default a UUID will be used
 	name string
@@ -262,6 +267,15 @@ func (d *DockerCompose) Down(ctx context.Context, opts ...StackDownOption) error
 		for cfg := range d.temporaryConfigs {
 			_ = os.Remove(cfg)
 		}
+	}()
+
+	// Close HTTP transport connections used by the compose CLI and the testcontainers
+	// Docker client to prevent net/http persistConn goroutine leaks.
+	defer func() {
+		if d.dockerCli != nil {
+			_ = d.dockerCli.Client().Close()
+		}
+		_ = d.dockerClient.Close()
 	}()
 
 	return d.composeService.Down(ctx, d.name, options.DownOptions)

--- a/modules/compose/compose_api.go
+++ b/modules/compose/compose_api.go
@@ -269,16 +269,33 @@ func (d *DockerCompose) Down(ctx context.Context, opts ...StackDownOption) error
 		}
 	}()
 
-	// Close HTTP transport connections used by the compose CLI and the testcontainers
-	// Docker client to prevent net/http persistConn goroutine leaks.
-	defer func() {
-		if d.dockerCli != nil {
-			_ = d.dockerCli.Client().Close()
-		}
-		_ = d.dockerClient.Close()
-	}()
-
 	return d.composeService.Down(ctx, d.name, options.DownOptions)
+}
+
+// Close releases the HTTP transport connections held by the internal Docker CLI
+// and the testcontainers Docker client, preventing net/http persistConn goroutine
+// leaks. Call Close after Down when the compose stack will no longer be used.
+func (d *DockerCompose) Close() error {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	var errs []error
+
+	if d.dockerCli != nil {
+		if cli := d.dockerCli.Client(); cli != nil {
+			if err := cli.Close(); err != nil {
+				errs = append(errs, fmt.Errorf("close docker cli client: %w", err))
+			}
+		}
+	}
+
+	if d.dockerClient != nil {
+		if err := d.dockerClient.Close(); err != nil {
+			errs = append(errs, fmt.Errorf("close docker client: %w", err))
+		}
+	}
+
+	return errors.Join(errs...)
 }
 
 func (d *DockerCompose) Up(ctx context.Context, opts ...StackUpOption) (err error) {

--- a/modules/compose/compose_api.go
+++ b/modules/compose/compose_api.go
@@ -287,12 +287,14 @@ func (d *DockerCompose) Close() error {
 				errs = append(errs, fmt.Errorf("close docker cli client: %w", err))
 			}
 		}
+		d.dockerCli = nil
 	}
 
 	if d.dockerClient != nil {
 		if err := d.dockerClient.Close(); err != nil {
 			errs = append(errs, fmt.Errorf("close docker client: %w", err))
 		}
+		d.dockerClient = nil
 	}
 
 	return errors.Join(errs...)

--- a/modules/compose/compose_goroutine_test.go
+++ b/modules/compose/compose_goroutine_test.go
@@ -1,0 +1,96 @@
+package compose
+
+import (
+	"context"
+	"fmt"
+	"runtime"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+// goroutineSnapshot captures the set of goroutine descriptions currently running,
+// keyed by their stack trace (without goroutine IDs so we can compare across calls).
+func goroutineSnapshot() map[string]int {
+	buf := make([]byte, 1<<20) // 1 MB
+	n := runtime.Stack(buf, true)
+	stacks := string(buf[:n])
+
+	counts := make(map[string]int)
+	// Each goroutine block starts with "goroutine N [..."
+	for _, block := range strings.Split(stacks, "\n\n") {
+		block = strings.TrimSpace(block)
+		if block == "" {
+			continue
+		}
+		// Strip the first line ("goroutine N [state]:") to make it ID-independent.
+		lines := strings.SplitN(block, "\n", 2)
+		if len(lines) < 2 {
+			continue
+		}
+		key := strings.TrimSpace(lines[1])
+		counts[key]++
+	}
+	return counts
+}
+
+// goroutineLeaks returns descriptions of goroutines that are present in after but not before,
+// excluding known background goroutines that are not related to the test.
+func goroutineLeaks(before, after map[string]int) []string {
+	var leaks []string
+	for stack, afterCount := range after {
+		beforeCount := before[stack]
+		if afterCount > beforeCount {
+			leaks = append(leaks, fmt.Sprintf("(+%d) %s", afterCount-beforeCount, stack))
+		}
+	}
+	return leaks
+}
+
+// TestDockerComposeGoroutineLeak verifies that calling Up() followed by Down() does not
+// leak net/http persistConn goroutines from the internal Docker CLI HTTP transport.
+//
+// Before the fix, two goroutines per Up+Down cycle were leaked permanently:
+//
+//	net/http.(*persistConn).readLoop
+//	net/http.(*persistConn).writeLoop
+func TestDockerComposeGoroutineLeak(t *testing.T) {
+	path, _ := RenderComposeSimple(t)
+
+	// Allow goroutines from previous tests to settle before taking baseline.
+	time.Sleep(200 * time.Millisecond)
+	before := goroutineSnapshot()
+
+	compose, err := NewDockerCompose(path)
+	require.NoError(t, err, "NewDockerCompose()")
+
+	ctx := context.Background()
+
+	err = compose.Up(ctx, Wait(true))
+	require.NoError(t, err, "compose.Up()")
+
+	err = compose.Down(ctx, RemoveOrphans(true), RemoveVolumes(true))
+	require.NoError(t, err, "compose.Down()")
+
+	// Allow goroutines to fully exit after Down().
+	time.Sleep(500 * time.Millisecond)
+
+	after := goroutineSnapshot()
+	leaks := goroutineLeaks(before, after)
+
+	// Filter to only HTTP transport goroutines that are the known leak.
+	var httpLeaks []string
+	for _, l := range leaks {
+		if strings.Contains(l, "persistConn") {
+			httpLeaks = append(httpLeaks, l)
+		}
+	}
+
+	assert.Empty(t, httpLeaks,
+		"net/http transport goroutines leaked after compose Down().\n"+
+			"These indicate the Docker CLI HTTP client was not closed.\n"+
+			"Leaked goroutines:\n%s", strings.Join(httpLeaks, "\n"))
+}

--- a/modules/compose/compose_goroutine_test.go
+++ b/modules/compose/compose_goroutine_test.go
@@ -2,95 +2,46 @@ package compose
 
 import (
 	"context"
-	"fmt"
-	"runtime"
-	"strings"
 	"testing"
-	"time"
 
-	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
+	"go.uber.org/goleak"
 )
 
-// goroutineSnapshot captures the set of goroutine descriptions currently running,
-// keyed by their stack trace (without goroutine IDs so we can compare across calls).
-func goroutineSnapshot() map[string]int {
-	buf := make([]byte, 1<<20) // 1 MB
-	n := runtime.Stack(buf, true)
-	stacks := string(buf[:n])
-
-	counts := make(map[string]int)
-	// Each goroutine block starts with "goroutine N [..."
-	for _, block := range strings.Split(stacks, "\n\n") {
-		block = strings.TrimSpace(block)
-		if block == "" {
-			continue
-		}
-		// Strip the first line ("goroutine N [state]:") to make it ID-independent.
-		lines := strings.SplitN(block, "\n", 2)
-		if len(lines) < 2 {
-			continue
-		}
-		key := strings.TrimSpace(lines[1])
-		counts[key]++
-	}
-	return counts
-}
-
-// goroutineLeaks returns descriptions of goroutines that are present in after but not before,
-// excluding known background goroutines that are not related to the test.
-func goroutineLeaks(before, after map[string]int) []string {
-	var leaks []string
-	for stack, afterCount := range after {
-		beforeCount := before[stack]
-		if afterCount > beforeCount {
-			leaks = append(leaks, fmt.Sprintf("(+%d) %s", afterCount-beforeCount, stack))
-		}
-	}
-	return leaks
-}
-
-// TestDockerComposeGoroutineLeak verifies that calling Up() followed by Down() does not
-// leak net/http persistConn goroutines from the internal Docker CLI HTTP transport.
+// TestDockerComposeGoroutineLeak verifies that calling Up() followed by Down()
+// and Close() does not leak net/http persistConn goroutines from the internal
+// Docker CLI HTTP transport.
 //
 // Before the fix, two goroutines per Up+Down cycle were leaked permanently:
 //
 //	net/http.(*persistConn).readLoop
 //	net/http.(*persistConn).writeLoop
+//
+// Note: Reaper (Ryuk) goroutines are intentionally excluded from this check;
+// they are a separate pre-existing issue tracked in the same issue but require
+// a distinct fix involving Reaper termination signal handling in Down().
 func TestDockerComposeGoroutineLeak(t *testing.T) {
 	path, _ := RenderComposeSimple(t)
-
-	// Allow goroutines from previous tests to settle before taking baseline.
-	time.Sleep(200 * time.Millisecond)
-	before := goroutineSnapshot()
 
 	compose, err := NewDockerCompose(path)
 	require.NoError(t, err, "NewDockerCompose()")
 
+	// Snapshot goroutines after NewDockerCompose establishes its initial
+	// Docker provider connection, so IgnoreCurrent covers the provider's
+	// keep-alive transport goroutines that pre-date the compose Up/Down cycle.
+	// We additionally ignore Reaper goroutines since they are tracked separately.
+	ignoreExisting := goleak.IgnoreCurrent()
+
+	t.Cleanup(func() {
+		require.NoError(t, compose.Close(), "compose.Close()")
+		goleak.VerifyNone(t,
+			ignoreExisting,
+			goleak.IgnoreTopFunction("github.com/testcontainers/testcontainers-go.(*Reaper).connect.func1"),
+		)
+	})
+
 	ctx := context.Background()
 
-	err = compose.Up(ctx, Wait(true))
-	require.NoError(t, err, "compose.Up()")
-
-	err = compose.Down(ctx, RemoveOrphans(true), RemoveVolumes(true))
-	require.NoError(t, err, "compose.Down()")
-
-	// Allow goroutines to fully exit after Down().
-	time.Sleep(500 * time.Millisecond)
-
-	after := goroutineSnapshot()
-	leaks := goroutineLeaks(before, after)
-
-	// Filter to only HTTP transport goroutines that are the known leak.
-	var httpLeaks []string
-	for _, l := range leaks {
-		if strings.Contains(l, "persistConn") {
-			httpLeaks = append(httpLeaks, l)
-		}
-	}
-
-	assert.Empty(t, httpLeaks,
-		"net/http transport goroutines leaked after compose Down().\n"+
-			"These indicate the Docker CLI HTTP client was not closed.\n"+
-			"Leaked goroutines:\n%s", strings.Join(httpLeaks, "\n"))
+	require.NoError(t, compose.Up(ctx, Wait(true)), "compose.Up()")
+	require.NoError(t, compose.Down(ctx, RemoveOrphans(true), RemoveVolumes(true)), "compose.Down()")
 }

--- a/modules/compose/compose_goroutine_test.go
+++ b/modules/compose/compose_goroutine_test.go
@@ -29,13 +29,16 @@ func TestDockerComposeGoroutineLeak(t *testing.T) {
 	// Snapshot goroutines after NewDockerCompose establishes its initial
 	// Docker provider connection, so IgnoreCurrent covers the provider's
 	// keep-alive transport goroutines that pre-date the compose Up/Down cycle.
-	// We additionally ignore Reaper goroutines since they are tracked separately.
 	ignoreExisting := goleak.IgnoreCurrent()
 
+	// Register Close cleanup first so it runs last (t.Cleanup is LIFO).
+	// goleak.VerifyNone is called here so it runs after both Down and Close.
 	t.Cleanup(func() {
 		require.NoError(t, compose.Close(), "compose.Close()")
 		goleak.VerifyNone(t,
 			ignoreExisting,
+			// TODO(#2008): Remove this ignore when the Reaper goroutine leak is fixed.
+			// This references an internal anonymous closure that may change if Reaper is refactored.
 			goleak.IgnoreTopFunction("github.com/testcontainers/testcontainers-go.(*Reaper).connect.func1"),
 		)
 	})
@@ -43,5 +46,10 @@ func TestDockerComposeGoroutineLeak(t *testing.T) {
 	ctx := context.Background()
 
 	require.NoError(t, compose.Up(ctx, Wait(true)), "compose.Up()")
-	require.NoError(t, compose.Down(ctx, RemoveOrphans(true), RemoveVolumes(true)), "compose.Down()")
+
+	// Register Down cleanup after Up so it runs before Close (t.Cleanup is LIFO).
+	// This ensures Down is called even if the test panics after Up succeeds.
+	t.Cleanup(func() {
+		require.NoError(t, compose.Down(ctx, RemoveOrphans(true), RemoveVolumes(true)), "compose.Down()")
+	})
 }

--- a/modules/compose/go.mod
+++ b/modules/compose/go.mod
@@ -13,6 +13,7 @@ require (
 	github.com/moby/moby/client v0.4.0
 	github.com/stretchr/testify v1.11.1
 	github.com/testcontainers/testcontainers-go v0.42.0
+	go.uber.org/goleak v1.3.0
 	go.yaml.in/yaml/v3 v3.0.4
 	golang.org/x/sync v0.20.0
 )

